File: java/L300/LeetCode0368.java
package L300;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class LeetCode0368 {
    public List<Integer> largestDivisibleSubset(int[] nums) { 
        List<Integer> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return res;
        }
        Arrays.sort(nums);
        int[][] dp = new int[nums.length][2];
        for (int i = 0;i < nums.length;++i) {
            dp[i][0] = 1;
            dp[i][1] = -1;
            for (int j = i - 1;j >= 0;--j) {
                if (nums[i] % nums[j] == 0) {
                    if (dp[j][0] + 1 > dp[i][0]) {
                        dp[i][0] = dp[j][0] + 1;
                        dp[i][1] = j;
                    }
                }
            }
        }
        for (int i = nums.length - 1;i >= 0;--i) {
            if (dp[i][0] > res.size()) {
                res.clear();
                for (int j = i;j != -1;j = dp[j][1]) {
                    res.add(nums[j]);
                }
            }
        }
        return res;
    }
}

File: cpp/L300/LeetCode0368.cc
#include <algorithm>
#include <vector>
using std::vector;

class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());  
        int N = nums.size();
        vector<vector<int>> dp(N, vector<int>(2, -1));
        int maxLen{0};
        for (int i = 0;i < N;++i) {
            dp[i][0] = 1;
            for (int j = i - 1;j >= 0;--j) {
                if (nums[i] % nums[j] == 0 && dp[j][0] + 1 > dp[i][0]) {
                    dp[i][0] = dp[j][0] + 1;
                    dp[i][1] = j;
                }
            }
            maxLen = std::max(maxLen, dp[i][0]);
        }
        vector<int> ans;
        for (int i = N - 1;i >= 0;--i) {
            if (maxLen != dp[i][0]) {
                continue;
            }
            int j = i;
            while (j >= 0) {
                ans.push_back(nums[j]);
                j = dp[j][1];
            }
            break;
        }
        return ans;
    }
};